xapian-letor/bin/trainletor.cc

```diff
@@ -0,0 +1,146 @@
+/** @file trainletor.cc
+ * @brief Command line tool to prepare Letor training file and train the model
+ */
+/* Copyright (C) 2004,2005,2006,2007,2008,2009,2010,2015 Olly Betts
+ * Copyright (C) 2011 Parth Gupta
+ * Copyright (C) 2016 Ayush Tomar
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include <xapian.h>
+#include <xapian-letor.h>
+
+#include <cstdlib>
+#include <cstring>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <map>
+#include <set>
+#include <math.h>
+
+#include "gnu_getopt.h"
+
+using namespace std;
+
+#define PROG_NAME "letor-train"
+#define PROG_DESC "Command line tool to prepare letor training file and train the model"
+
+static void show_usage() {
+    cout << "Usage: " PROG_NAME " [OPTIONS]\n"
+"Options:\n"
+"  --db=DIRECTORY  database to search (multiple databases may be specified)\n"
+"  --query=PATH  path to query file\n"
+"  --qrel=PATH  path to qrel file\n"
+"  --msize=MSIZE   maximum number of matches to return\n"
+"  --train=PATH  path to file by which training file is to be saved\n"
+"  --model=PATH  path to file by training file is to be saved\n"
+"  --help          display this help and exit\n"
+"  --version       output version information and exit\n";
+}
+
+int
+main(int argc, char **argv)
+try {
+    const char * opts = "a:b:c:d:e:f:g:h";
```

Comment:
It's probably harder to get familiar with short options (which is what I tend to use with most CLI programs) if they don't "connect" with their use at all.

This might be easier if we split creating the training file out from training the model. Then we could have:

```
$ xapian-prepare-trainingfile [options] <queryfile> <qrelfile> <trainingfile>
$ xapian-train-model [options] <trainingfile> <modelfile>
```

or something. Since most of the options are required for each stage, they may well be better as arguments instead.